applications/GeoMechanicsApplication/custom_strategies/schemes/backward_euler_quasistatic_U_Pw_scheme.hpp

```diff
@@ -42,31 +43,45 @@ class BackwardEulerQuasistaticUPwScheme : public NewmarkQuasistaticUPwScheme<TSp
         KRATOS_TRY
 
         this->SetDeltaTime(rModelPart.GetProcessInfo()[DELTA_TIME]);
-        rModelPart.GetProcessInfo()[VELOCITY_COEFFICIENT]    = 1.0/this->GetDeltaTime();
-        rModelPart.GetProcessInfo()[DT_PRESSURE_COEFFICIENT] = 1.0/this->GetDeltaTime();
+        rModelPart.GetProcessInfo()[VELOCITY_COEFFICIENT] = 1.0 / this->GetDeltaTime();
+        rModelPart.GetProcessInfo()[DT_PRESSURE_COEFFICIENT] = 1.0 / this->GetDeltaTime();
 
         KRATOS_CATCH("")
     }
 
+    template <class T>
+    void SetDerivative(const Variable<T>& derivative_variable,
+                       const Variable<T>& instance_variable,
+                       Node& rNode)
+    {
+        rNode.FastGetSolutionStepValue(derivative_variable) =
+            (rNode.FastGetSolutionStepValue(instance_variable, 0) -
+             rNode.FastGetSolutionStepValue(instance_variable, 1)) /
+            this->GetDeltaTime();
+    }
+
     inline void UpdateVariablesDerivatives(ModelPart& rModelPart) override
     {
         KRATOS_TRY
 
-        //Update Acceleration, Velocity and DtPressure
-        block_for_each(rModelPart.Nodes(), [this](Node& rNode) {
-            // refactor, extract the (a -b)/mDeltaTime that happens 3 times here
-            noalias(rNode.FastGetSolutionStepValue(VELOCITY))     = (  rNode.FastGetSolutionStepValue(DISPLACEMENT)
-                                                                                   - rNode.FastGetSolutionStepValue(DISPLACEMENT, 1)) / this->GetDeltaTime();
-
-            noalias(rNode.FastGetSolutionStepValue(ACCELERATION)) = (  rNode.FastGetSolutionStepValue(VELOCITY)
-                                                                                   - rNode.FastGetSolutionStepValue(VELOCITY,1) ) / this->GetDeltaTime();
-
-            rNode.FastGetSolutionStepValue(DT_WATER_PRESSURE)        = (  rNode.FastGetSolutionStepValue(WATER_PRESSURE)
-                                                                                  - rNode.FastGetSolutionStepValue(WATER_PRESSURE, 1)) / this->GetDeltaTime();
-
+        // Update Acceleration, Velocity and DtPressure
+        block_for_each(rModelPart.Nodes(), [this](Node& rNode)
+        {
+            for (const auto& r_variable_with_derivative : this->GetVariableDerivatives())
+            {
+                SetDerivative(r_variable_with_derivative.first_time_derivative,
+                              r_variable_with_derivative.instance, rNode);
+
+                // Make sure that setting the second_time_derivative is done
+                // after setting the first_time_derivative.
+                SetDerivative(r_variable_with_derivative.second_time_derivative,
+                              r_variable_with_derivative.first_time_derivative, rNode);
+            }
+
+            SetDerivative(DT_WATER_PRESSURE, WATER_PRESSURE, rNode);
```

Comment:
WATER_PRESSURE at this moment is already stored in the baseclass. However, the problem now is that the BackwardEulerQuasistaticUPwScheme still derives from the GeneralizedNewmarkScheme instead of the BackwardEulerScheme (see snippet right). We have made efforts as you see to improve the structure, but we didn't fix everything yet (to not get gigantic PRs). I would propose to fix this as part of our next refactoring effort (I created a draft issue for it on the backlog).
![image](https://github.com/KratosMultiphysics/Kratos/assets/56549273/39507dc7-7da9-4660-b298-bc5ea380b6e3)

Reply:
Thank you for the clarification. Sure it can be done later.